src/subscription/secure-3d/secure-3d-modal/Secure3dModal.jsx

```diff
@@ -0,0 +1,158 @@
+import React, { useState, useRef, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import { useDispatch, useSelector } from 'react-redux';
+
+import {
+  ModalDialog,
+} from '@edx/paragon';
+import {
+  PaymentElement,
+} from '@stripe/react-stripe-js';
+
+import { subscriptionStatusSelector } from '../../data/status/selectors';
+import { detailsSelector } from '../../data/details/selectors';
+
+import {
+  onSuccessful3DS,
+} from '../../data/status/actions';
+
+/**
+ * Secure3DModal
+ * This modal implements the 3DS functionality for
+ * both trial and non-trial purchases. It uses the setupIntent
+ * for trial and paymentIntent for non-trial
+ * purchases in order to load 3DS details inside an iFrame modal
+ */
+export const Secure3DModal = ({ stripe, elements }) => {
+  const dispatch = useDispatch();
+  const modalRef = useRef('inActive');
+  const [isOpen, setOpen] = useState(false);
+
+  const { isTrialEligible } = useSelector(detailsSelector);
+  const { status, confirmationClientSecret } = useSelector(subscriptionStatusSelector);
+
+  /**
+   * loadSecureDetails
+  */
+  const loadSecureDetails = (url) => {
+    setTimeout(() => {
+      const container = document.getElementById('secure-3ds-wrapper');
+      const iframe = document.createElement('iframe');
+      iframe.src = url;
+      iframe.width = '100%';
+      iframe.height = '100%';
+      container.appendChild(iframe);
+    });
+  };
+
+  /**
+   * retrieveSetupIntent
+   * call this method with subscription in trial mode
+   */
+  const retrieveSetupIntent = async () => {
+    const response = await stripe.retrieveSetupIntent(confirmationClientSecret);
+    if (response.error) { throw response.error; }
+    return response.setupIntent;
+  };
+
+  /**
+   * retrievePaymentIntent
+   * call this method with subscription in non-trial mode
+   */
+  const retrievePaymentIntent = async () => {
+    const response = await stripe.retrievePaymentIntent(confirmationClientSecret);
+    if (response.error) { throw response.error; }
+    return response.paymentIntent;
+  };
+
+  /**
+   * on3DSComplete
+   * callback to handle 3DS completion state
+   */
+  const on3DSComplete = async () => {
+    // Hide the 3DS UI
+    if (modalRef.current === 'active') {
+      // completed 3DS, set to inActive so multiple state updates doesn't trigger the 3DS update
+      modalRef.current = 'inActive';
```

Comment:
Do you need the camel-casing here vs. just `'inactive'`?

Reply:
updated.